### PR TITLE
Prevent creating empty sessions

### PR DESCRIPTION
**Pull request by zeripath**
Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: modules/setting/session.go [hunk 1]
```diff
@@ -30,6 +34,16 @@ func newSessionService() {
 	SessionConfig.Secure = Cfg.Section("session").Key("COOKIE_SECURE").MustBool(false)
 	SessionConfig.Gclifetime = Cfg.Section("session").Key("GC_INTERVAL_TIME").MustInt64(86400)
 	SessionConfig.Maxlifetime = Cfg.Section("session").Key("SESSION_LIFE_TIME").MustInt64(86400)
+	emptySessionsAreVirtual := Cfg.Section("session").Key("AVOID_CREATING_EMPTY_SESSIONS").MustBool(true)
```
Comment (lunny) [inline modules/setting/session.go]: Why this is optional? I think it must be true.

**Review by lunny — CHANGES_REQUESTED** [review]

File: modules/setting/session.go [hunk 2]
```diff
@@ -5,11 +5,15 @@
 package setting
 
 import (
+	"encoding/json"
 	"path"
 	"path/filepath"
 	"strings"
 
 	"code.gitea.io/gitea/modules/log"
+	// This ensures that OnlyCookieIfNotLoggedInProvider is available
```
Comment (zeripath) [inline modules/setting/session.go]:
```suggestion
	// This ensures that VirtualSessionProvider is available
```

File: modules/setting/session.go (hunk 1 above)
Comment (zeripath) [reply modules/setting/session.go]: I guess VirtualSessionProvider changes the semantics of Count. 

I guess I just didn't want to enforce this completely.

**Comment by lunny** [comment]
Could you add some integration tests to confirm the unlogin page didn't create session but logined page did?

**Comment by zeripath** [comment]
@lunny it seems that's not an easy thing to do. Our integration tests run with the memory session provider which does not appear to be interrogatable.

**Comment by zeripath** [comment]
Tests are up.

File: integrations/create_no_session_test.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
```
Comment (techknowlogick) [inline integrations/create_no_session_test.go]: Tiny nitpick, but could you update year?

Comment (zeripath) [reply integrations/create_no_session_test.go]:
```suggestion
// Copyright 2019 The Gitea Authors. All rights reserved.
```

**Comment by zeripath** [comment]
Tests now working.

**Review by techknowlogick — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
@zeripath thanks

**Pull request merged**